pytest_mitmproxy.py

```diff
@@ -1 +1,101 @@
 # -*- coding: utf-8 -*-
+import re
+import os
+import six
+import time
+import errno
+import shutil
+import pytest
+from contextlib import contextmanager
+from tempfile import TemporaryFile
+if six.PY3:
+    from urllib.parse import urlparse
+    from subprocess import Popen, CalledProcessError, TimeoutExpired
+elif six.PY2:
+    from urlparse import urlparse
+    from subprocess32 import Popen, CalledProcessError
+
+# clean the dumps each run
+if os.path.isdir("traffic_dumps"):
+    shutil.rmtree("traffic_dumps")
+
+
+def pytest_addoption(parser):
+    parser.addoption(
+        "--mitmproxy-filter",
+        action="store",
+        default=None,
+        help="filter specification for mitmproxy"
+    )
+
+
+@contextmanager
+def new_proxy(name, scope, flt=None):
+    ''' Start a new mitmdump process and
+    return the address for clients to connect
+
+    The captured content will be written to
+    ./<dump_dir>/<scope>/<name>
+    '''
+    dump_dir = "traffic_dumps"
+
+    try:
+        os.makedirs(os.path.join(dump_dir, scope), mode=0o755)
+    except OSError as err:
+        if err.errno != errno.EEXIST:
+            raise
+
+    dump_file = os.path.join(dump_dir, scope, name)
+
+    args = ["mitmdump", "--port=0", "--wfile={}".format(dump_file)]
+    if flt:
+        args.append(flt)
+
+    def read_all(fp):
+        fp.seek(0)
+        return fp.read().decode("utf-8")
+
+    with TemporaryFile() as stdout:
+        ps = Popen(args, stdout=stdout)
+        # Wait for status line
+        for _ in range(50):
```

Comment:
No.

Yes pipe will always be readable. But that's not the point of select and friends. Sockets are also always readable, etc.

The point is that it'll sleep the process until there's something *available* to read - as in at least one read operation can happen without blocking. Meaning at least once write operation has happened on the remote side.

Since stdout is line buffered by default, its very reasonable to then assume at least one line of output should be in the pipe after polling.